docs/site/Context.md

```diff
@@ -530,3 +530,101 @@ class MyController {
   }
 }
 ```
+
+## Configuration by convention
+
+To allow bound items in the context to be configured, we introduce some
+conventions and corresponding APIs to make it simple and consistent.
+
+We treat configurations for bound items in the context as dependencies, which
+can be resolved and injected in the same way of other forms of dependencies. For
+example, the `RestServer` can be configured with `RestServerConfig`.
+
+Let's first look at an example:
+
+```ts
+export class RestServer {
+  constructor(
+    @inject(CoreBindings.APPLICATION_INSTANCE) app: Application,
+    @inject(RestBindings.CONFIG, {optional: true})
+    config: RestServerConfig = {},
+  ) {
+    // ...
+  }
+  // ...
+}
+```
+
+The configuration (`RestServerConfig`) itself is a binding
+(`RestBindings.CONFIG`) in the context. It's independent of the binding for
+`RestServer`. The caveat is that we need to maintain a different binding key for
+the configuration. Referencing a hard-coded key for the configuration also makes
+it impossible to have more than one instances of the `RestServer` to be
+configured with different options, such as `protocol` or `port`.
+
+To solve these problems, we introduce an accompanying binding for an item that
+expects configuration. For example:
+
+- `servers.RestServer.server1`: RestServer
+- `servers.RestServer.server1.$config`: RestServerConfig
+
+- `servers.RestServer.server2`: RestServer
+- `servers.RestServer.server2.$config`: RestServerConfig
+
+The following APIs are available to enforce/leverage this convention:
+
+1. ctx.configure('servers.RestServer.server1') => Binding for the configuration
```

Comment:
I am concerned that in order to contribute configuration, the code needs a reference to the target Context instance. It creates a coupling that makes it difficult for extensions to contribute configuration bindings.

Can you add API for creating standalone configuration bindings (no context instance required)? Perhaps a short-cut leveraging `configBindingKeyFor` helper?

For example, we can add a static method `Binding.configure` that will create a new `Binding` instance with the correct configuration key:

```ts
// in a component, export the following binding:
class MyComponent {
  bindings: [
    // some other bindings

    Binding.configure('servers.RestServer.server1').to({port: 0})
  ],
});

// in mountComponent - already in place
if (component.bindings) {
  for (const binding of component.bindings) {
    app.add(binding);
  }
}
```

I am not sure if `Binding.configure` is the right name, feel free to pick a better one!

I am also fine to leave out such new helper out of scope of this pull request, as long as the documentation explains how to create standalone Binding instances for configuration (e.g. using `configBindingKeyFor`)

Reply:
Good idea. I'll add `Binding.configure`.

Reply:
> Good idea. I'll add `Binding.configure`

Please describe the new API in the documentation too. (E.g. add it to the list here.)